pydeseq2/dds.py

```diff
@@ -218,55 +239,190 @@ def __init__(
+        self.continuous_factors = (
+            continuous_factors if continuous_factors is not None else []
         )
-        self.continuous_factors = continuous_factors
 
-        if self.obs[self.design_factors].isna().any().any():
-            raise ValueError("NaNs are not allowed in the design factors.")
-        self.obs[self.design_factors] = self.obs[self.design_factors].astype(str)
+        # If it's a list we convert it right back to a formulaic string
+        if isinstance(self.design_factors, list):
+            # Little bit of careful formating as below to remove space around : symbols
+            self.design_factors = [
+                re.sub(r"\ \:|\:\ ", ":", factor) for factor in self.design_factors
+            ]
+            # We extract all single factors
+            is_interaction_terms = [":" in factor for factor in self.design_factors]
+            extracted_interacting_terms_single_factors = list(
+                itertools.chain(
+                    *[
+                        factor.split(":")
+                        for is_interaction, factor in zip(
+                            is_interaction_terms, self.design_factors
+                        )
+                        if is_interaction
+                    ]
+                )
+            )
+            extracted_single_factors = [
+                factor
+                for is_interaction, factor in zip(
+                    is_interaction_terms, self.design_factors
+                )
+                if not is_interaction
+            ]
+            extracted_single_factors = list(
+                set(extracted_single_factors)
+                & set(extracted_interacting_terms_single_factors)
+            )
+            self.single_design_factors = extracted_single_factors
+            self.design_factors = "~" + "+".join(self.design_factors)
+
+        # Following lines handle the case where the user provides a formula
+        # of the type formulaic can recognize or a column everything will be
+        # converted to formulaic formulas as in the first case
+        elif isinstance(self.design_factors, str):
+            # TODO uncomment when we support all of formulaic grammar
+            # try:
+            #     Formula(self.design_factors)
+            # except FormulaInvalidError as err:
+            #     raise (
+            #         f"Design factor {self.design_factors} is not a valid formula"
+            #     ) from err
+            # Remove potential spaces around +,: or ~ signs for easier parsing
+            # of individual factor
+            if not self.design_factors.startswith("~"):
+                self.design_factors = "~" + self.design_factors
+            self.design_factors = re.sub(r"\ \+|\+\ ", "+", self.design_factors)
+            self.design_factors = re.sub(r"\ \:|\:\ ", ":", self.design_factors)
+            self.design_factors = re.sub(r"\ \~|\~\ ", ":", self.design_factors)
```

Comment:
Why not at least support `A*B` which is equivalent of `A + B + A:B`? 

What's the limiting factor of supporting any formulaic formula, as ultimately, it will just return a design matrix anyway?